src/test/parse.ts

```diff
@@ -114,6 +121,74 @@ export function parseErrorsFromMarkup(text: string): LintError[] {
     return lintErrors;
 }
 
+/**
+ * Process `message` as follows:
+ * - search `substitutions` for an exact match and return the substitution
+ * - try to format the message when it looks like: name % ('substitution1' [, "substitution2" [, ...]])
+ * - or return it unchanged
+ */
+function substituteMessage(substitutions: Map<string, string>, message: string): string {
+    const substitution = substitutions.get(message);
+    if (substitution !== undefined) {
+        return substitution;
+    }
+    return formatMessage(substitutions, message);
+}
+
+/**
+ * Tries to format the message when it has the correct format or returned unchanged:  name % ('substitution1' [, "substitution2" [, ...]])
+ * Where `name` is the name of a message substitution that is used as format string.
+ * If `name` is not found in `substitutions`, `message` is returned unchanged.
+ */
+function formatMessage(substitutions: Map<string, string>, message: string): string {
```

Comment:
The reason I suggested the name change, is that if one were to look at just this function signature, you'd assume that `substitutions: Map<string, string>` is a map between placeholders and replacement values, and that `message` is the template string.

Also, I think it's fine calling these templates since template strings don't necessarily have placeholders